source/s3_meta_request.c

```diff
@@ -842,20 +832,269 @@ static void s_s3_meta_request_send_request_finish_default(
+            (void *)request->meta_request,
+            (void *)request,
+            aws_last_error(),
+            aws_error_str(aws_last_error()));
+
+        goto error_ref_clean_up;
+    }
+
+    return;
+
+error_ref_clean_up:
+
+    aws_s3_request_release(request);
+    aws_s3_meta_request_finish(meta_request, NULL, 0, aws_last_error());
+    aws_s3_meta_request_release(meta_request);
+}
+
+void aws_s3_meta_request_handle_error(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_request *request,
+    int error_code) {
+    AWS_PRECONDITION(meta_request);
+
+    AWS_LOGF_ERROR(
+        AWS_LS_S3_META_REQUEST,
+        "id=%p Meta request handling error %d (%s). (request=%p)",
+        (void *)meta_request,
+        error_code,
+        aws_error_str(error_code),
+        (void *)request);
+
+    int response_status = 0;
+
+    if (request != NULL) {
+        request->send_data.error_code = error_code;
+        response_status = request->send_data.response_status;
+    }
+
+    /* If the request is NULL or we have a response that indicates retrying would be futile, then bail on the entire
+     * meta request. */
+    if (request == NULL || error_code == AWS_ERROR_S3_INVALID_RESPONSE_STATUS) {
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request cannot recover from error %d (%s). (request=%p, response status=%d)",
+            (void *)meta_request,
+            error_code,
+            aws_error_str(error_code),
+            (void *)request,
+            response_status);
+
+        aws_s3_meta_request_finish(meta_request, request, response_status, error_code);
+        return;
+    }
+
+    struct aws_s3_client *client = aws_s3_meta_request_acquire_client(meta_request);
+
+    /* If we were able to get the point of handling an error, the client should still be around. */
+    AWS_ASSERT(client);
+    AWS_ASSERT(request);
+    AWS_ASSERT(request->meta_request == meta_request);
+
+    aws_s3_request_acquire(request);
+    aws_s3_meta_request_acquire(meta_request);
+
+    /* If the retry token is NULL, try to grab one, otherwise, just re-use the token from the request's last retry. */
+    if (request->retry_token == NULL) {
+        s_s3_meta_request_acquire_retry_token(client->retry_strategy, request);
+    } else {
+        s_s3_meta_request_queue_retry_with_token(
+            client->retry_strategy, AWS_ERROR_SUCCESS, request->retry_token, request);
+    }
+
+    /* Release the reference to the client that we got with aws_s3_meta_request_get_client. */
+    aws_s3_client_release(client);
+}
+
+void aws_s3_meta_request_retry_queue_push(struct aws_s3_meta_request *meta_request, struct aws_s3_request *request) {
+    AWS_PRECONDITION(meta_request);
+    AWS_PRECONDITION(request->meta_request == meta_request);
 
     aws_s3_request_acquire(request);
 
     aws_s3_meta_request_lock_synced_data(meta_request);
 
+    aws_s3_meta_request_release(request->meta_request);
```

Comment:
What makes this unlink/link on retry necessary from a ref handling perspective?  Is it impossible to initiate cancel if the retry queue also has refs to the meta request?

Reply:
The idea was that once the meta request has the request in its queue, it shouldn't prevent clean up of the meta request by keeping a reference to it.

In the refactor, we're actually changing to keep the reference, but just clear out the queue on finish.